akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -210,6 +210,169 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
+      if (activeSubstreamsMap.isEmpty || (!hasNextElement && !activeSubstreamsMap.values.exists(_.firstPush))) {
+        activeSubstreamsMap.values.foreach(_.substreamSource.complete())
+        completeStage()
+        true
+      } else false
+
+    private def fail(ex: Throwable): Unit = {
+      activeSubstreamsMap.values.foreach(_.substreamSource.fail(ex))
+      failStage(ex)
+    }
+
+    private def needToPull: Boolean = !(hasBeenPulled(in) || isClosed(in) || hasNextElement)
+
+    override def preStart(): Unit = timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
+
+    setHandlers(in, out, new OutHandler with InHandler {
+
+      override def onPull(): Unit = {
+        substreamPushed match {
+          case Some(substreamHandler) ⇒
+            push(out, Source.fromGraph(substreamHandler.substreamSource.source))
+            scheduleOnce(substreamHandler.key, timeout)
+            substreamPushed = None
+          case None ⇒
+            if (hasNextElement) {
+              activeSubstreamsMap(nextElement._1).substreamSource.push(nextElement._2)
+              nextElement = null
+            } else tryPull(in)
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = fail(ex)
+
+      override def onDownstreamFinish(): Unit =
+        if (activeSubstreamsMap.isEmpty) completeStage() else setKeepGoing(true)
+
+      override def onPush(): Unit = {
+        val elem = grab(in)
+        val keyOption = getKeyOption(elem)
+
+        keyOption match {
+          case Some(key) ⇒ activeSubstreamsMap.get(key) match {
+            case Some(handler) ⇒
+              if (handler.substreamSource.isAvailable) handler.substreamSource.push(elem)
+              else nextElement = key -> elem
+            case None ⇒
+              if (activeSubstreamsMap.size == maxSubstreams)
+                fail(new IllegalStateException(s"Cannot open substream for key '$key': too many substreams open"))
+              else if (closedSubstreams(key) && !hasBeenPulled(in)) pull(in)
+              else runSubstream(key, elem)
+          }
+          case None ⇒ // do nothing
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (!completeWholeStage()) setKeepGoing(true)
+      }
+    })
+
+    private def runSubstream(key: K, value: T): Unit = {
+      val substreamSource = new SubSourceOutlet[T]("GroupBySource " + nextId)
+      val handler = new SubstreamHandler(key, value, substreamSource)
+
+      substreamSource.setHandler(handler)
+      activeSubstreamsMap += key -> handler
+
+      if (isAvailable(out)) {
+        push(out, Source.fromGraph(substreamSource.source))
+        scheduleOnce(key, timeout)
+        substreamPushed = None
+      } else {
+        setKeepGoing(true)
+        substreamPushed = Some(handler)
+      }
+    }
+
+    private def getKeyOption(elem: T): Option[K] = {
+      try {
+        Some(f(elem))
```

Comment:
Two thoughts: 
1. do we really want to allocate a Some for each element that passes by?
2. if we do want options, shouldn't this be `Option(f(elem))` or contain a `null` check of the return value?